command/acl_role_update.go

```diff
@@ -0,0 +1,217 @@
+    characters.
+
+  -policy-name
+    Specifies a policy to associate with the role identified by their name. This
+    flag can be specified multiple times.
+
+  -no-merge
+    Do not merge the current role information with what is provided to the
+    command. Instead overwrite all fields with the exception of the role ID
+    which is immutable.
+
+  -json
+    Output the ACL role in a JSON format.
+
+  -t
+    Format and display the ACL role using a Go template.
+`
+
+	return strings.TrimSpace(helpText)
+}
+
+func (a *ACLRoleUpdateCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(a.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-name":        complete.PredictAnything,
+			"-description": complete.PredictAnything,
+			"-no-merge":    complete.PredictNothing,
+			"-policy-name": complete.PredictAnything,
+			"-json":        complete.PredictNothing,
+			"-t":           complete.PredictAnything,
+		})
+}
+
+func (a *ACLRoleUpdateCommand) AutocompleteArgs() complete.Predictor { return complete.PredictNothing }
+
+// Synopsis satisfies the cli.Command Synopsis function.
+func (a *ACLRoleUpdateCommand) Synopsis() string { return "Update an existing ACL role" }
+
+// Name returns the name of this command.
+func (*ACLRoleUpdateCommand) Name() string { return "acl role update" }
+
+// Run satisfies the cli.Command Run function.
+func (a *ACLRoleUpdateCommand) Run(args []string) int {
+
+	flags := a.Meta.FlagSet(a.Name(), FlagSetClient)
+	flags.Usage = func() { a.Ui.Output(a.Help()) }
+	flags.StringVar(&a.name, "name", "", "")
+	flags.StringVar(&a.description, "description", "", "")
+	flags.Var((funcVar)(func(s string) error {
+		a.policyNames = append(a.policyNames, s)
+		return nil
+	}), "policy-name", "")
+	flags.BoolVar(&a.noMerge, "no-merge", false, "")
+	flags.BoolVar(&a.json, "json", false, "")
+	flags.StringVar(&a.tmpl, "t", "", "")
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	// Check that we got exactly one argument which is expected to be the ACL
+	// role ID.
+	if len(flags.Args()) != 1 {
+		a.Ui.Error("This command takes one argument: <acl_role_id>")
+		a.Ui.Error(commandErrorText(a))
+		return 1
+	}
+
+	// Get the HTTP client.
+	client, err := a.Meta.Client()
+	if err != nil {
+		a.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	aclRoleID := flags.Args()[0]
+
+	// Read the current role in both cases, so we can fail better if not found.
+	currentRole, _, err := client.ACLRoles().Get(aclRoleID, nil)
+	if err != nil {
+		a.Ui.Error(fmt.Sprintf("Error when retrieving ACL role: %v", err))
+		return 1
+	}
+
+	var updatedRole api.ACLRole
+
+	// Depending on whether we are merging or not, we need to take a different
+	// approach.
+	switch a.noMerge {
+	case true:
+
+		// Perform some basic validation on the submitted role information to
+		// avoid sending API and RPC requests which will fail basic validation.
+		if a.name == "" {
```

Comment:
Should we validate max length. or are we relying on the API/RPC to bubble that up? I guess that applies to other commands as well.

Reply:
I am following what we do for policies for the most part, however, where we do validation is something I always question because it seems there is a little bit here and there. Generally though, I think it would need to be in the HTTP API rather than the CLI so we would cover both CLI and API users in the validation.

I'll resolve this for now, but we should chat internally and come up with a plan around how this should look at how we can incrementally implement it.